pep-0646.rst

```diff
@@ -743,11 +753,220 @@ is available in `cpython/23527`_. A preliminary version of the version
+
+::
+
+    def matrix_matrix_muliply(x: Array[K, N], Array[N, M]) -> Array[K, M]: ...
+    def matrix_vector_multiply(x: Array[K, N], Array[N]) -> Array[K]: ...
+    
+Here, all axes sharing the label ``N`` would be constrained to have the same size within
+a given signature, but there would be no relationship between ``N`` in *different* signatures.
+Such a local scoping rule would be important to avoid forcing, say, all functions within one
+module using a label ``Batch`` to use the same batch size.
+
+The disadvantage of this approach is that we have no ability to enforce shape properties across
+different calls. For example, we can't address the problem mentioned in `Motivation`_: if
+one function returns an array with leading dimensions 'Time × Batch', and another function
+takes the same array assuming leading dimensions 'Batch × Time', we have no way of detecting this.
+(Even allowing for broader scoping rules would not completely address the problem. For example,
+we might have a special kind of label that referred to the same size throughout the whole program.
+But then, what if ``Time`` and ``Batch`` were the same size?)
+
+The main advantage is that in some cases, axis sizes really are what we care about. This is true
+for both simple linear algebra operations such as the matrix manipulations above, but also in more
+complicated transformations such as convolutional layers in neural networks, where it would be of
+great utility to the programmer to be able to inspect the array size after each layer using
+static analysis.
+
+Names as Semantic Identity
+''''''''''''''''''''''''''
+
+A second approach (the one that most of the examples in this PEP are based around)
+would be to have names constrain axis *type* but not axis *size*.
+
+This would enable us to solve the problem of enforcing shape properties across calls.
+For example:
+
+::
+
+    # lib.py
+    
+    class Batch: pass
+    class Time: pass
+    
+    def make_array() -> Array[Batch, Time]: ...
+    
+    # user.py
+    
+    from lib import Batch, Time
+    
+    # `Batch` and `Time` have the same identity as in `lib`,
+    # so must take array as produced by `lib.make_array`
+    def use_array(x: Array[Batch, Time]): ...
+    
+In many cases, this is the more important thing to verify; we care more about
+which axis is which than what the specific size of each axis is.
+
+It also does not preclude use cases where we wish to describe shape transformations
+where we don't know semantic identity ahead of time. For example, we can write:
+
+::
+
+    K = TypeVar('K')
+    N = TypeVar('N')
+
+    def matrix_vector_multiply(x: Array[K, N], Array[N]) -> Array[K]: ...
+    
+We can then use this with:
+    
+    class Batch: pass
+    class Values: pass
+    
+    batch_of_values: Array[Batch, Values]
+    value_weights: Array[Values]
+    matrix_vector_multiply(batch_of_values, value_weights)
+    # Result is Array[Batch]
+
```

Comment:
I like the idea of reorganising around use cases, though I still think it's important to have some discussion of what the 'scope' of names is, since this is something that JAX folks specifically asked for (and I could see being easily confusing for other folks too). Where do you think the best place for that would be? 

> Names as Local Sizes: Already doable using Python generics.

It is?

Reply:
> Names as Local Sizes: Already doable using Python generics.
> It is?

From your example and from shoyer's comment below, it looks like it:

> (Note: from our offline discussions, I'm pretty sure this is supported with type variables, but I'd still like to see it in the PEP for the sake of clarification for maintainers of array libraries.)


> I still think it's important to have some discussion of what the 'scope' of names is, since this is something that JAX folks specifically asked for (and I could see being easily confusing for other folks too). Where do you think the best place for that would be?

The "Meaning and Scope of Axis Names" section had a lot of questions, which made me more confused by the end. How about mentioning the meaning and scope of the name in each use case? That would answer questions people had.

Reply:
Right, I've rewritten this section around use cases instead, and indeed it does seem a lot clearer. Good suggestion :) What do you think of it now?

Reply:
Looks great! Thanks for taking the time.